### ISSUE TITLE
Vagrant support

### ISSUE THREAD

**Issue by BSFishy** [gitpod-io/gitpod#1251]
I have been using Gitpod for a little bit now, because it lets me write code on any machine I use while I'm at school. This is great, because I always have an environment where I can write code without worrying about clutter, dependencies, or whatever else. While this is all fine and good, I also want my projects to be easily accessible to anyone, even those not using Gitpod. To help with this process, in terms of dependencies and whatnot, I came across Vagrant. It has, in fact, been exactly what I've been wanting for a while now, because I've wanted a Docker-esque system that is targeted for development. Vagrant, in theory, could eliminate the friction of getting the dependencies needed for my project, no matter the platform or setup. I am interested in integrating my projects with it, but am somewhat skeptical about how it would work using Gitpod.

I use Docker files in all of my projects to install dependencies, so I know how all of that works. My curiosity comes in with using a VM inside of Gitpod. Is that even supported and/or recommended? I don't exactly know the specs of the machines running the containers, so I don't know how running an entire VM could affect the performance/storage.

Furthermore, I would like to raise Vagrant as an alternative instead of Docker inside of Gitpod. I'll admit that I don't entirely know how the backend works, but I think it would be great to be able to just define a single Vagrant file and have independent users as well as Gitpod users use that file, rather than having Gitpod users using a Docker file and independent users trying to figure out dependencies on their own, or using a separate Vagrant file that is potentially out of sync with the Docker file.

**Comment by Kreyren** [comment]
Can you give me a summary of that? Maybe all you need is to add `RUN apt install vagrant` in dockerfile ?

**Comment by BSFishy** [comment]
[Vagrant](https://www.vagrantup.com/) is a tool to easily manage virtual machines, aimed towards software development. It is a flexible system built to work using other software such as VMWare or VirtualBox. The idea is that you can clone a repository and run `vagrant up` to setup and start a virtual machine that is configured using a `Vagrantfile` in the repository's root. Then you can run `vagrant ssh` to use that virtual machine's terminal.

While I don't doubt the possibility of installing and using it in Gitpod, I am more concerned with using unnecessary resources associated with setting up and starting an entirely new virtual machine within a container. It would essentially be a completely new OS installation, running alongside the already existing OS used by Gitpod. That is where my concern stems from, and is why I think it could be useful to incorporate Vagrant as an alternative to Dockerfiles.

**Comment by Kreyren** [comment]
FWIW: this configuration https://github.com/Kreyren/gitpod-tests1/tree/9b8020e48b5ebdbe7ba4ef4be9368e9be76ce8cd results in 

```
gitpod /workspace/gitpod-tests1 $ VBoxManage --version
WARNING: The character device /dev/vboxdrv does not exist.
         Please install the virtualbox-dkms package and the appropriate
         headers, most likely linux-headers-4.14.138+.

         You will not be able to start VMs until this problem is fixed.
6.0.6_Ubuntur129722
```

**Comment by Kreyren** [comment]
Apparently it is possible to pass `--device` argument in docker run that should give us the required access.

Checking weather it is sane to use

**Comment by JesterOrNot** [comment]
Hello @BSFishy and welcome to Gitpod! 🎉 
I've heard of vagrant, however I have never used it myself does it require KVM?

**Comment by Kreyren** [comment]
>  I've heard of vagrant, however, I have never used it myself does it require KVM?

vagrant is just a command line tool, the providers it uses are where the work is actually done. essentially vagrant captures the config you define and calls out to tools like vboxmanage or docker to do any work - From #vagrant on freenode

Seems that the virtual box is an issue here where we need /dev/vboxdrv as mensioned in https://github.com/gitpod-io/gitpod/issues/1251#issuecomment-589754601 to fix this issue

**Comment by JesterOrNot** [comment]
cc @geropl @csweichel @meysholdt

**Comment by Kreyren** [comment]
FWIW i'm in #vbox waiting for response

**Comment by Kreyren** [comment]
may be relevant: https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2008-3431

**Comment by Kreyren** [comment]
Relevant: https://www.scmagazine.com/home/security-news/oracles-virtualbox-vulnerability-leaked-by-disgruntled-researcher/

> “The E1000 has a vulnerability allowing an attacker with root/administrator privileges in a guest to escape to a host ring3.” Zelenyuk wrote in a technical write-up posted to his GitHub account in technical write-up. ”Then the attacker can use existing techniques to escalate privileges to ring 0 via /dev/vboxdrv.

**Comment by BSFishy** [comment]
> Also, it's up to the whole team but I would not like to see Vagrant replace Docker

I just wanted to clarify: I didn't mean to imply that Vagrant should outright replace Docker. What I was trying to say is that there could be an option in the config file to specify that the repository uses a Vagrantfile, which would be used to setup the container, rather than a Dockerfile.

**Comment by JesterOrNot** [comment]
I see you mean for users to have the option to write their config in a Vagrant or Dockerfile?

**Comment by JesterOrNot** [comment]
@BSFishy would running Vagrant in a workspace require root or Docker see #52

**Comment by Kreyren** [comment]
afaiu this expects the ability to run `vagrant up` which reads `Vagrantfile` to set up VMs for development, i.e testing the software on freebsd kernel.

**Comment by BSFishy** [comment]
> BSFishy would running Vagrant in a workspace require root or Docker see #52

I wouldn't assume so. I haven't actually worked with Vagrant yet, as I haven't had access to a development environment outside of Gitpod, but I think it only manipulates the underlying providers, i.e. VMWare or VirtualBox. If the provider needs root access to create or interact with virtual machines, it probably would too.

**Comment by JesterOrNot** [comment]
I know but would the VMs that vagrant is spinning up require root

**Comment by BSFishy** [comment]
In order to install packages through something like apt, I think so. Although since it is in a separate virtual environment, I would expect it to be fine...

**Comment by JesterOrNot** [comment]
This depends on #39 being fixed as we don't allow root inside our workspaces for security reasons @csweichel can explain it better than I can

**Comment by Kreyren** [comment]
Based on #vagrant it should work on rootless system, investigating atm

**Comment by BSFishy** [comment]
Just read through a tiny bit of VirtualBox documentation. From what I saw, actually creating and running virtual machines doesn't actually require root permissions. That would mean that I could technically install Vagrant and VirtualBox in a Gitpod container right now (ignoring the issue with installing VirtualBox). I could then create a Vagrantfile and spin up a virtual machine using Vagrant, where I would have root privilege.

**Comment by BSFishy** [comment]
**Breaking:** Vagrant actually has native support for [using Docker a its provider](https://www.vagrantup.com/docs/docker/). Potentially this could be used instead of using virtual machines? Although being able to make virtual machines in Gitpod would be very cool.

Additionally, Vagrant also allows you to [make your own provider](https://www.vagrantup.com/docs/plugins/providers.html). Potentially that could be used to better integrate it into Gitpod if the team does decide to go that route, or even just to be able to use it in the workspaces.

**Comment by Kreyren** [comment]
> Breaking: Vagrant actually has native support for using Docker a its provider. Potentially this could be used instead of using virtual machines?

Vagrant allows configuration of docker images this does not mean that it can run be used as invidual sandboxing method.

I'm currently having minor success with vagrant on Docker using KVM/QEMU which does not have the mensioned security issue.

**Comment by Kreyren** [comment]
cc @meysholdt @svenefftinge this configuration https://github.com/Kreyren/gitpod-tests1/tree/809456e4bf16825911b1e5f4e1695d9c210f8466 should work on gitpod assuming kvm module providing /dev/kvm.

**Comment by JesterOrNot** [comment]
We cannot use KVM @Kreyren

**Comment by JesterOrNot** [comment]
I'll start some testing in https://github.com/JesterOrNot/Gitpod-Vagrant

**Comment by Kreyren** [comment]
> We cannot use KVM @Kreyren

Elaborate

**Comment by JesterOrNot** [comment]
From our Slack channel
![image](https://user-images.githubusercontent.com/50386709/75083806-33b38800-54e1-11ea-9477-e1fa018901ca.png)

**Comment by Kreyren** [comment]
@JesterOrNot can you elaborate further? The hardware supports it based on available info

> Forcing me to use slack

Grrr

**Comment by JesterOrNot** [comment]
You don't have to use slack that is an internal TypeFox chat the issue that has more info has become private

**Comment by Kreyren** [comment]
Make it public then so that we can evaluate the options :p

**Comment by JesterOrNot** [comment]
I can't I don't even have access to it anymore 😅

**Comment by Kreyren** [comment]
@JesterOrNot Get the access then? Allowing KVM shoudn't be an issue in terms of security and maintainance we have the resources to configure it.

**Comment by Kreyren** [comment]
btw.  we also have the resources to run rootless docker which is based on available info a vulnerability waiting to hapend where i wasn't able to find any legitimate usecase for it to run as root.

**Comment by JesterOrNot** [comment]
https://blog.trendmicro.com/trendlabs-security-intelligence/why-running-a-privileged-container-in-docker-is-a-bad-idea/

**Comment by Kreyren** [comment]
@JesterOrNot From quickread i dont see anything that would conflict with proposed since i was told that running root on a rootless dockerd is sane assuming sane configuration.

**Comment by JesterOrNot** [comment]
![image](https://user-images.githubusercontent.com/50386709/75084174-eab10300-54e3-11ea-8d27-468cd4d63da3.png)

**Comment by Kreyren** [comment]
@JesterOrNot 
![image](https://user-images.githubusercontent.com/11302521/75084186-b3ac1300-551e-11ea-9016-dcab407f7370.png)

That's why i mensioned sane configuration (linux's permission system is very good compared to XNU if that's where the concern comes from), trying to fetch security checklist for you atm.

**Comment by Kreyren** [comment]
Referencing https://github.com/vagrant-libvirt/vagrant-libvirt/issues/1078#issuecomment-590098120 for more info.

**Comment by Kreyren** [comment]
Referencing https://github.com/docker/compose/issues/7244#issuecomment-589941607 for rootless docker.

**Comment by BSFishy** [comment]
Afaik, this issue relies more on the support of virtual machines in general, rather than the support of Vagrant. Should another issue be created to document VM support in general?

Although it would be nice for native support of Vagrant like that of Docker.

**Comment by BSFishy** [comment]
I've been working quite a bit, trying to get this to work these past two days. As @Kreyren said, the kvm kernel module needs to be loaded. To my knowledge, I think that is the last thing that needs to happen to allow all of this to work (although I very much could be wrong).

I have brought my repository [here](https://github.com/bitmodo/assertions) to a point where it could be working, but I am getting an error:

```text
Error while creating domain: Error saving the server: Call to virDomainDefineXML failed: invalid argument: could not find capabilities for domaintype=kvm
```

This indicates that the kernel module is not loaded. Again, I could be wrong that this is the last thing that needs to happen, but it would be great if we could have the option to enable this module.

**Comment by BSFishy** [comment]
@JesterOrNot So would it be possible to have the KVM kernel module enabled? Or is there some other outlying problem that would prevent that? I would like to be able to use Vagrant as soon as possible, so having the kernel module loaded would be the next step I believe.

**Comment by JesterOrNot** [comment]
This has been solved with the new root support see the following example which uses the docker backend! https://github.com/JesterOrNot/Gitpod-Vagrant Remember to have the feature preview option enabled from your settings as this solution leverages both the new root feature and the new docker support.

**Comment by jankeromnes** [comment]
Awesome! Many thanks @JesterOrNot! 🙏

Closing issue as per https://github.com/gitpod-io/gitpod/issues/1251#issuecomment-740675506

**Comment by kasperk81** [comment]
@JesterOrNot, I just tested on gitpod.io with:

```
sudo apt update && sudo apt install -y vagrant
mkdir foo && cd foo
vagrant init generic/alpine38
vagrant up
```
errors out:
> Bringing machine 'default' up with 'libvirt' provider...
Error while connecting to libvirt: Error making a connection to libvirt URI qemu:///system?no_verify=1&keyfile=/home/gitpod/.ssh/id_rsa:
Call to virConnectOpen failed: Failed to connect socket to '/var/run/libvirt/libvirt-sock': No such file or directory

see https://community.gitpod.io/t/cant-find-featurepreview-option-in-gitpod-settings/3428/4, there is no preview feature that we can enable under Settings anymore.
